Assets/MRTK/Core/Utilities/Physics/MixedRealityRaycaster.cs

```diff
@@ -19,96 +26,99 @@ public static bool RaycastSimplePhysicsStep(RayStep step, LayerMask[] prioritize
             return RaycastSimplePhysicsStep(step, step.Length, prioritizedLayerMasks, focusIndividualCompoundCollider, out physicsHit);
         }
 
+        private static readonly ProfilerMarker RaycastSimplePhysicsStepPerfMarker = new ProfilerMarker("[MRTK] MixedRealityRaycaster.RaycastSimplePhysicsStep");
+
         /// <summary>
         /// Simple raycasts each physics <see cref="Microsoft.MixedReality.Toolkit.Physics.RayStep"/> within a specified maximum distance.
         /// </summary>
         /// <returns>Whether or not the raycast hit something.</returns>
         public static bool RaycastSimplePhysicsStep(RayStep step, float maxDistance, LayerMask[] prioritizedLayerMasks, bool focusIndividualCompoundCollider, out RaycastHit physicsHit)
         {
-            Profiler.BeginSample("[MRTK] MixedRealityRaycaster.RaycastSimplePhysicsStep");
-
-            Debug.Assert(maxDistance > 0, "Length must be longer than zero!");
-            Debug.Assert(step.Direction != Vector3.zero, "Invalid step direction!");
-
-            bool result = prioritizedLayerMasks.Length == 1
-                // If there is only one priority, don't prioritize
-                ? UnityEngine.Physics.Raycast(step.Origin, step.Direction, out physicsHit, maxDistance, prioritizedLayerMasks[0])
-                // Raycast across all layers and prioritize
-                : TryGetPrioritizedPhysicsHit(UnityEngine.Physics.RaycastAll(step.Origin, step.Direction, maxDistance, UnityEngine.Physics.AllLayers), prioritizedLayerMasks, focusIndividualCompoundCollider, out physicsHit);
-
-            Profiler.EndSample(); // RaycastSimplePhysicsStep
-
-            return result;
+            using (RaycastSimplePhysicsStepPerfMarker.Auto())
+            {
+                Debug.Assert(maxDistance > 0, "Length must be longer than zero!");
+                Debug.Assert(step.Direction != Vector3.zero, "Invalid step direction!");
+
+                int hitCount = UnityEngine.Physics.RaycastNonAlloc(
+                    step.Origin,
+                    step.Direction,
+                    RaycastHits,
+                    maxDistance,
+                    (prioritizedLayerMasks.Length == 1) ? (int)prioritizedLayerMasks[0] : UnityEngine.Physics.AllLayers);
```

Comment:
I wonder if there's a perf gain to be had by only raycasting against the layer masks in `prioritizedLayerMasks` instead of `UnityEngine.Physics.AllLayers`?

Reply:
not sure, but at least the raycaster should ignore irrelevant layers. should investigate in a separate pass